org.eclipse.xtext/src/org/eclipse/xtext/resource/persistence/SerializableResourceDescription.java

```diff
@@ -0,0 +1,165 @@
+/**
+ * Copyright (c) 2014, 2020 itemis AG (http://www.itemis.eu) and others.
+ * This program and the accompanying materials are made available under the
+ * terms of the Eclipse Public License 2.0 which is available at
+ * http://www.eclipse.org/legal/epl-2.0.
+ * 
+ * SPDX-License-Identifier: EPL-2.0
+ */
+package org.eclipse.xtext.resource.persistence;
+
+import static org.eclipse.xtext.resource.persistence.SerializationExtensions.*;
+import static org.eclipse.xtext.xbase.lib.IterableExtensions.*;
+
+import java.io.Externalizable;
+import java.io.IOException;
+import java.io.ObjectInput;
+import java.io.ObjectOutput;
+import java.util.ArrayList;
+import java.util.Collections;
+import java.util.HashMap;
+import java.util.List;
+
+import org.eclipse.emf.common.util.URI;
+import org.eclipse.xtext.naming.QualifiedName;
+import org.eclipse.xtext.resource.IEObjectDescription;
+import org.eclipse.xtext.resource.IReferenceDescription;
+import org.eclipse.xtext.resource.IResourceDescription;
+import org.eclipse.xtext.resource.impl.AbstractResourceDescription;
+
+import com.google.common.collect.Lists;
+
+/**
+ * @author Sven Efftinge - Initial contribution and API
+ * 
+ * @since 2.8
+ */
+public class SerializableResourceDescription extends AbstractResourceDescription implements Externalizable {
+
+	private List<SerializableEObjectDescription> descriptions = Collections.emptyList();
+
+	private List<SerializableReferenceDescription> references = Collections.emptyList();
+
+	private List<QualifiedName> importedNames = Collections.emptyList();
+
+	private URI uRI;
+
+	public void updateResourceURI(URI uri) {
+		for (SerializableReferenceDescription ref : references)
+			ref.updateResourceURI(uri, uRI);
+		for (SerializableEObjectDescription desc : descriptions)
+			desc.updateResourceURI(uri);
+		uRI = uri;
+	}
+
+	@Override
+	@SuppressWarnings("unchecked")
+	protected List<IEObjectDescription> computeExportedObjects() {
+		return ((List<IEObjectDescription>) ((List<?>) this.descriptions));
```

Comment:
I don't immediately say why the casts are needed? Isn't SerializableEObjectDescription instanceof IEObjectDescription

Reply:
True. But a `List< SerializableEObjectDescription>` is not a `List<IEObjectDescription>`.